platform/security/idp/security-idp-plugin/security-idp-plugin-requestsubjectvalidator/src/main/java/org/codice/ddf/security/idp/plugin/requestsubjectvalidator/RequestSubjectValidatorPresignPlugin.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.idp.plugin.requestsubjectvalidator;
+
+import ddf.security.samlp.SamlProtocol.Binding;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.opensaml.core.xml.schema.XSString;
+import org.opensaml.saml.common.SAMLRuntimeException;
+import org.opensaml.saml.saml2.core.Assertion;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.NameID;
+import org.opensaml.saml.saml2.core.Response;
+import org.opensaml.saml.saml2.core.Subject;
+
+/**
+ * This plugin is responsible for verifying that the {@code Subject} of the received
+ * {@code AuthnRequest} refers to the same principle as the {@code Subject} of the {@code Response}.
+ */
+public class RequestSubjectValidatorPresignPlugin implements SamlPresignPlugin {
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+
+    // short circuits for when no Subject/NameID value is on the AuthnRequest
+    Subject authnRequestSubject = authnRequest.getSubject();
+    if (authnRequestSubject == null) {
+      return;
+    }
+
+    NameID authnRequestNameId = authnRequestSubject.getNameID();
+    if (authnRequestNameId == null) {
+      return;
+    }
+
+    String authnRequestId = authnRequestNameId.getValue();
+    if (StringUtils.isBlank(authnRequestNameId.getValue())) {
+      return;
+    }
+
+    if (!authnRequestSubject.getSubjectConfirmations().isEmpty()) {
+      throw new SAMLRuntimeException(
+          "The AuthnRequest contained a Subject with both a NameID and "
+              + "1 or more SubjectConfirmations");
```

Comment:
Good catch on the spec requirement 🎉 
✏️ I think this message is a little misleading because it seems to say that it'd be okay to not have a NameID and have subject confirmations. Because the spec is so explicit, I think this message shouldn't say anything about the NameID and the check should be put ahead any checks for the NameID.

Reply:
Great catch! I couldn't find it in the core document, but in profiles it says:
```
Note that the service provider MAY include a <Subject> element in the request that names the actual identity about which it wishes to receive an assertion. This element MUST NOT contain any <SubjectConfirmation> elements.
```